Automatically replace spaces in parameter names with underscores

**AS A** Service Creator  
**I WANT** spaces in parameter names to be automatically replaced with underscores  
**SO THAT** variable names remain valid and do not cause execution errors

---

### Acceptance Criteria

- [x] When the user types a **space** in a parameter/variable name field, it is **immediately replaced with `_`**
  - Example: typing `user name` becomes `user_name`
- [x] Manual pasting of names containing spaces also triggers auto-replacement
- [x] Behavior applies consistently in:
  - [x] “Määra” element  
  - [x] API parameter mapping fields 
  - [x] Any element that defines new parameters
- [x] Exported JSON always contains valid identifiers without spaces
- [x] No breaking errors when executing in Ruuter or evaluating JS expressions

---

##Additional info:
```
Caused by: javax.script.ScriptException: ReferenceError: asd is not defined in <eval> at line number 1 at column number 1
	at com.oracle.truffle.js.scriptengine.GraalJSScriptEngine.toScriptException(GraalJSScriptEngine.java:519)
	at com.oracle.truffle.js.scriptengine.GraalJSScriptEngine.eval(GraalJSScriptEngine.java:487)
	at com.oracle.truffle.js.scriptengine.GraalJSScriptEngine.eval(GraalJSScriptEngine.java:453)
	at java.scripting/javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:231)
	at ee.buerokratt.ruuter.helper.ScriptingHelper.evaluate(ScriptingHelper.java:150)
	... 99 common frames omitted
Caused by: org.graalvm.polyglot.PolyglotException: ReferenceError: asd is not defined
	at <js>.:program(<eval>:1)
	at org.graalvm.polyglot.Context.eval(Context.java:403)
	at com.oracle.truffle.js.scriptengine.GraalJSScriptEngine.eval(GraalJSScriptEngine.java:485)
```

--- a/GUI/src/components/ApiEndpointCard/Endpoints/RequestVariables/VariableCell/index.tsx
+++ b/GUI/src/components/ApiEndpointCard/Endpoints/RequestVariables/VariableCell/index.tsx
@@ -23,9 +23,14 @@ const VariableCell: React.FC<VariableCellProps> = ({ row, updateRowVariable, var
       name={`endpoint-variable-${row.id}`}
       label=""
       onChange={(e) => {
-        onValueChange(row.id, e.target.value);
-        setInputValue(e.target.value);
-        updateRowVariable(row.id, e.target.value);
+        const value = e.target.value.trimStart().replaceAll(/_+/g, '_');
+        const hasSpecialCharacters = /[^\p{L}\p{N}_ ]/u;
+        if (!hasSpecialCharacters.test(value) && !value.startsWith(' ')) {
+          const formattedValue = value.replaceAll(' ', '_');
+          onValueChange(row.id, formattedValue);
+          setInputValue(formattedValue);
+          updateRowVariable(row.id, formattedValue);
+        }
       }}
       value={inputValue}
       placeholder={t('newService.endpoint.variable') + '..'}

--- a/GUI/src/components/FlowElementsPopup/AssignBuilder/AssignElement.tsx
+++ b/GUI/src/components/FlowElementsPopup/AssignBuilder/AssignElement.tsx
@@ -45,7 +45,11 @@ const AssignElement: React.FC<AssignElementProps> = ({
   const [isObjectEditorOpen, setIsObjectEditorOpen] = useState(element.isObject ?? false);
 
   const changeKey = (e: React.ChangeEvent<HTMLInputElement>) => {
-    onChange({ ...element, key: e.target.value });
+    const value = e.target.value.trimStart().replaceAll(/_+/g, '_');
+    const hasSpecialCharacters = /[^\p{L}\p{N}_ ]/u;
+    if (!hasSpecialCharacters.test(value) && !value.startsWith(' ')) {
+      onChange({ ...element, key: value.replaceAll(' ', '_') });
+    }
   };
 
   const changeValue = (e: React.ChangeEvent<HTMLInputElement>) => {

--- a/GUI/src/components/FlowElementsPopup/RuleBuilder/ruleElement.tsx
+++ b/GUI/src/components/FlowElementsPopup/RuleBuilder/ruleElement.tsx
@@ -16,7 +16,15 @@ interface RuleElementProps {
 }
 
 const RuleElement: React.FC<RuleElementProps> = ({ rule, onRemove, onChange }) => {
-  const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
+  const handleFieldChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
+    const value = e.target.value.trimStart().replaceAll(/_+/g, '_');
+    const hasSpecialCharacters = /[^\p{L}\p{N}_ ]/u;
+    if (!hasSpecialCharacters.test(value) && !value.startsWith(' ')) {
+      change(e.target.name, value.replaceAll(' ', '_'));
+    }
+  };
+
+  const handleValueChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
     change(e.target.name, e.target.value);
   };
 
@@ -82,7 +90,7 @@ const RuleElement: React.FC<RuleElementProps> = ({ rule, onRemove, onChange }) =
           <FormInput
             value={rule.field}
             name="field"
-            onChange={handleChange}
+            onChange={handleFieldChange}
             onDrop={handleFieldDrop}
             label=""
             hideLabel
@@ -109,7 +117,7 @@ const RuleElement: React.FC<RuleElementProps> = ({ rule, onRemove, onChange }) =
           <FormInput
             value={rule.value}
             name="value"
-            onChange={handleChange}
+            onChange={handleValueChange}
             onDrop={handleValueDrop}
             label=""
             hideLabel

--- a/GUI/src/components/FlowElementsPopup/index.tsx
+++ b/GUI/src/components/FlowElementsPopup/index.tsx
@@ -236,6 +236,10 @@ const FlowElementsPopup: React.FC = () => {
     if (stepType === StepType.Assign) {
       const flatEndpointVariables = endpointsVariables.map((endpoint) => endpoint.chips).flat();
       assignElements.forEach((element) => {
+
+        const key = removeTrailingUnderscores(element.key);
+        element.key = key;
+        
         // Convert simple values such as "some input" to simple string
         if (!isTemplate(element.value)) {
           element.value = stringToTemplate('"' + element.value + '"');

--- a/GUI/src/components/FlowElementsPopup/index.tsx
+++ b/GUI/src/components/FlowElementsPopup/index.tsx
@@ -35,7 +35,7 @@ import TextfieldTestContent from './TextfieldTestContent';
 import { servicesRequestsExplain } from '../../resources/api-constants';
 import api from '../../services/api-dev';
 import { StepType } from '../../types';
-import { getInitialGroup } from './RuleBuilder/types';
+import { getInitialGroup, GroupOrRule } from './RuleBuilder/types';
 import './styles.scss';
 
 const FlowElementsPopup: React.FC = () => {
@@ -198,6 +198,17 @@ const FlowElementsPopup: React.FC = () => {
 
     if (stepType === StepType.Input || stepType === StepType.Condition) {
       const rulesArray = Array.isArray(rules) ? rules : [];
+      for (const item of rulesArray) {
+        const processRuleField = (obj: GroupOrRule) => {
+          if ('field' in obj) obj.field = removeTrailingUnderscores(obj.field);
+          else {
+            for (const child of obj.children) {
+              processRuleField(child);
+            }
+          }
+        };
+        processRuleField(item);
+      }
       updatedNode.data.rules = node.data.rules
         ? { ...node.data.rules, children: rulesArray }
         : { ...getInitialGroup(), children: rulesArray };

--- a/GUI/src/components/FlowElementsPopup/index.tsx
+++ b/GUI/src/components/FlowElementsPopup/index.tsx
@@ -197,21 +197,7 @@ const FlowElementsPopup: React.FC = () => {
     };
 
     if (stepType === StepType.Input || stepType === StepType.Condition) {
-      const rulesArray = Array.isArray(rules) ? rules : [];
-      for (const item of rulesArray) {
-        const processRuleField = (obj: GroupOrRule) => {
-          if ('field' in obj) obj.field = removeTrailingUnderscores(obj.field);
-          else {
-            for (const child of obj.children) {
-              processRuleField(child);
-            }
-          }
-        };
-        processRuleField(item);
-      }
-      updatedNode.data.rules = node.data.rules
-        ? { ...node.data.rules, children: rulesArray }
-        : { ...getInitialGroup(), children: rulesArray };
+      prepareRulesForSaving(updatedNode);
     }
 
     if (stepType === StepType.MultiChoiceQuestion) {
@@ -219,70 +205,100 @@ const FlowElementsPopup: React.FC = () => {
     }
 
     if (stepType === StepType.UserDefined) {
-      if (updatedNode.data?.endpoint?.definitions) {
-        for (const definition of updatedNode.data.endpoint.definitions) {
-          for (const section of ['body', 'headers', 'params'] as const) {
-            if (definition[section]?.variables) {
-              for (const v of definition[section].variables) {
-                v.name = removeTrailingUnderscores(v.name);
-              }
-            }
-          }
-        }
-      }
-      const newLabel = updatedNode.data.label?.toString().split(' ');
-      if (updatedNode.data.endpoint?.name) {
-        newLabel[0] = updatedNode.data.endpoint?.name ?? node.data.label?.toString().split(' ')[0];
-        const nodeWithSameLabel = instance
-          ?.getNodes()
-          .find((n) => n.data.label === newLabel.join(' ') && n.id !== updatedNode.id);
-        if (nodeWithSameLabel) {
-          useToastStore.getState().error({
-            title: t('newService.toast.elementNameAlreadyExists'),
-            message: t('newService.toast.elementNameAlreadyExistsMessage'),
-          });
-          return;
-        }
-        updatedNode.data.label = newLabel.join(' ');
-      }
-      useServiceStore.getState().loadEndpointsResponseVariables();
+      prepareEndpointsForSaving(updatedNode);
     }
 
     if (stepType === StepType.Assign) {
-      const flatEndpointVariables = endpointsVariables.map((endpoint) => endpoint.chips).flat();
-      assignElements.forEach((element) => {
-        const key = removeTrailingUnderscores(element.key);
-        element.key = key;
-
-        // Convert simple values such as "some input" to simple string
-        if (!isTemplate(element.value)) {
-          element.value = stringToTemplate('"' + element.value + '"');
-          return;
-        }
-
-        const fullPath = templateToString(element.value);
-        const endpointVariable = flatEndpointVariables.find((variable) => fullPath.startsWith(String(variable.value)));
-
-        if (!endpointVariable) {
-          // Element is not an object so no data for ObjectTree
-          return;
-        }
-
-        const value = String(endpointVariable.value);
-        const remainingPath = fullPath.substring(
-          fullPath[value.length] === '['
-            ? // Uses array notation, e.g. endpointVariable[1].something; needed for backwards compatibility
-              value.length
-            : // Uses object notation, e.g. endpointVariable.1.something
-              value.length + 1,
-        );
-        element.data = remainingPath ? getValueByPath(endpointVariable.data, remainingPath) : endpointVariable.data;
-      });
-      updatedNode.data.assignElements = assignElements;
+      prepareAssignForSaving(updatedNode);
     }
 
     useServiceStore.getState().handlePopupSave(updatedNode);
     onClose();
+  };
+
+  const prepareAssignForSaving = (updatedNode: Node<NodeDataProps>) => {
+    const flatEndpointVariables = endpointsVariables.map((endpoint) => endpoint.chips).flat();
+    assignElements.forEach((element) => {
+      const key = removeTrailingUnderscores(element.key);
+      element.key = key;
+
+      // Convert simple values such as "some input" to simple string
+      if (!isTemplate(element.value)) {
+        element.value = stringToTemplate('"' + element.value + '"');
+        return;
+      }
+
+      const fullPath = templateToString(element.value);
+      const endpointVariable = flatEndpointVariables.find((variable) => fullPath.startsWith(String(variable.value)));
+
+      if (!endpointVariable) {
+        // Element is not an object so no data for ObjectTree
+        return;
+      }
+
+      const value = String(endpointVariable.value);
+      const remainingPath = fullPath.substring(
+        fullPath[value.length] === '['
+          ? // Uses array notation, e.g. endpointVariable[1].something; needed for backwards compatibility
+            value.length
+          : // Uses object notation, e.g. endpointVariable.1.something
+            value.length + 1,
+      );
+      element.data = remainingPath ? getValueByPath(endpointVariable.data, remainingPath) : endpointVariable.data;
+    });
+    updatedNode.data.assignElements = assignElements;
+  };
+
+  const prepareEndpointsForSaving = (updatedNode: Node<NodeDataProps>) => {
+    filterOutEndpointsTrailingUnderscores(updatedNode);
+    const newLabel = updatedNode.data.label?.toString().split(' ');
+    if (updatedNode.data.endpoint?.name) {
+      newLabel[0] = updatedNode.data.endpoint?.name ?? node.data.label?.toString().split(' ')[0];
+      const nodeWithSameLabel = instance
+        ?.getNodes()
+        .find((n) => n.data.label === newLabel.join(' ') && n.id !== updatedNode.id);
+      if (nodeWithSameLabel) {
+        useToastStore.getState().error({
+          title: t('newService.toast.elementNameAlreadyExists'),
+          message: t('newService.toast.elementNameAlreadyExistsMessage'),
+        });
+        return;
+      }
+      updatedNode.data.label = newLabel.join(' ');
+    }
+    useServiceStore.getState().loadEndpointsResponseVariables();
+  };
+
+  const filterOutEndpointsTrailingUnderscores = (updatedNode: Node<NodeDataProps>) => {
+    if (updatedNode.data?.endpoint?.definitions) {
+      for (const definition of updatedNode.data.endpoint.definitions) {
+        for (const section of ['body', 'headers', 'params'] as const) {
+          if (definition[section]?.variables) {
+            for (const v of definition[section].variables) {
+              v.name = removeTrailingUnderscores(v.name);
+            }
+          }
+        }
+      }
+    }
+  };
+
+  const prepareRulesForSaving = (updatedNode: Node<NodeDataProps>) => {
+    const rulesArray = Array.isArray(rules) ? rules : [];
+    for (const item of rulesArray) {
+      const processRuleField = (obj: GroupOrRule) => {
+        if ('field' in obj) obj.field = removeTrailingUnderscores(obj.field);
+        else {
+          for (const child of obj.children) {
+            processRuleField(child);
+          }
+        }
+      };
+      processRuleField(item);
+    }
+    updatedNode.data.rules = node.data.rules
+      ? { ...node.data.rules, children: rulesArray }
+      : { ...getInitialGroup(), children: rulesArray };
   };
 
   const handleJsonRequestClick = async () => {

--- a/GUI/src/components/FlowElementsPopup/index.tsx
+++ b/GUI/src/components/FlowElementsPopup/index.tsx
@@ -215,6 +215,17 @@ const FlowElementsPopup: React.FC = () => {
     }
 
     if (stepType === StepType.UserDefined) {
+      if (updatedNode.data?.endpoint?.definitions) {
+        for (const definition of updatedNode.data.endpoint.definitions) {
+          for (const section of ['body', 'headers', 'params'] as const) {
+            if (definition[section]?.variables) {
+              for (const v of definition[section].variables) {
+                v.name = removeTrailingUnderscores(v.name);
+              }
+            }
+          }
+        }
+      }
       const newLabel = updatedNode.data.label?.toString().split(' ');
       if (updatedNode.data.endpoint?.name) {
         newLabel[0] = updatedNode.data.endpoint?.name ?? node.data.label?.toString().split(' ')[0];

--- a/GUI/src/components/FlowElementsPopup/index.tsx
+++ b/GUI/src/components/FlowElementsPopup/index.tsx
@@ -251,10 +251,9 @@ const FlowElementsPopup: React.FC = () => {
     if (stepType === StepType.Assign) {
       const flatEndpointVariables = endpointsVariables.map((endpoint) => endpoint.chips).flat();
       assignElements.forEach((element) => {
-
         const key = removeTrailingUnderscores(element.key);
         element.key = key;
-        
+
         // Convert simple values such as "some input" to simple string
         if (!isTemplate(element.value)) {
           element.value = stringToTemplate('"' + element.value + '"');

--- a/GUI/src/services/service-builder.ts
+++ b/GUI/src/services/service-builder.ts
@@ -34,17 +34,7 @@ export async function saveEndpoints(endpoints: EndpointData[], onSuccess?: () =>
   }
 
   endpoints.forEach((endpoint) => {
-    if (endpoint.definitions) {
-      for (const definition of endpoint.definitions) {
-        for (const section of ['body', 'headers', 'params'] as const) {
-          if (definition[section]?.variables) {
-            for (const v of definition[section].variables) {
-              v.name = removeTrailingUnderscores(v.name);
-            }
-          }
-        }
-      }
-    }
+    filterOutEndpointsTrailingUnderscores(endpoint);
     if (endpoint.isNew) {
       tasks.push(createEndpointAndUpdateState(endpoint));
     } else {
@@ -59,6 +49,20 @@ export async function saveEndpoints(endpoints: EndpointData[], onSuccess?: () =>
   });
 
   await Promise.all(tasks).then(onSuccess).catch(onError);
+}
+
+function filterOutEndpointsTrailingUnderscores(endpoint: EndpointData) {
+  if (endpoint.definitions) {
+    for (const definition of endpoint.definitions) {
+      for (const section of ['body', 'headers', 'params'] as const) {
+        if (definition[section]?.variables) {
+          for (const v of definition[section].variables) {
+            v.name = removeTrailingUnderscores(v.name);
+          }
+        }
+      }
+    }
+  }
 }
 
 async function createEndpointAndUpdateState(endpoint: EndpointData): Promise<any> {

--- a/GUI/src/services/service-builder.ts
+++ b/GUI/src/services/service-builder.ts
@@ -34,6 +34,17 @@ export async function saveEndpoints(endpoints: EndpointData[], onSuccess?: () =>
   }
 
   endpoints.forEach((endpoint) => {
+    if (endpoint.definitions) {
+      for (const definition of endpoint.definitions) {
+        for (const section of ['body', 'headers', 'params'] as const) {
+          if (definition[section]?.variables) {
+            for (const v of definition[section].variables) {
+              v.name = removeTrailingUnderscores(v.name);
+            }
+          }
+        }
+      }
+    }
     if (endpoint.isNew) {
       tasks.push(createEndpointAndUpdateState(endpoint));
     } else {

